docs: i18n code snippets

petebacondarwin opened the pull request:
Fixes #43215

petebacondarwin committed:
build(docs-infra): support docregions in `.sh` example files

This commit tells the docregion parser how to process `.sh` files.

petebacondarwin committed:
docs: move shell snippets to external example files

mary-poppins commented:
You can preview 44983ce at https://pr43216-44983ce.ngbuilds.io/.

mary-poppins commented:
You can preview 13b5093 at https://pr43216-13b5093.ngbuilds.io/.

petebacondarwin committed:
docs: move global locale import to external example file

petebacondarwin committed:
docs: move plural function example to external file

petebacondarwin committed:
build(docs-infra): support docregions in `.conf` files

This commit tells the docregion parser how to process files that end in `.conf`.

petebacondarwin committed:
docs: move nginx and apache config blocks to external files

petebacondarwin committed:
docs: move duplicate custom id snippets to external examples

mary-poppins commented:
You can preview f60a8e7 at https://pr43216-f60a8e7.ngbuilds.io/.

mary-poppins commented:
You can preview 5dd512d at https://pr43216-5dd512d.ngbuilds.io/.

petebacondarwin committed:
build(docs-infra): switch JSON examples to use inline docregion comments

Now JSON files should be commented using

```
// #docregion
```

rather than

```
/* #docregion */
```

There are currently no JSON files with either kind of comments so this change
has no impact yet.

But when we add some examples with such comments, they will also benefit
from `// ... ` style docplasters to join docregion fragments.

petebacondarwin committed:
build(docs-infra): indent docplaster markers to the level of the docregion

This change makes docplaster markers fit better with the surrounding code
that is being rendered.  Instead of:

```
{
  "a": 100,
// ...
  "b": 200,
// ...
}
```

you can now get

```
{
  "a": 100,
  // ...
  "b": 200,
  // ...
}
```

petebacondarwin committed:
docs: move i18n angular.json snippets to external file

Note that because `angular.json` is a "boilerplate" file we cannot
just add comments to the one that is actually used in the application.

Instead this commit makes a copy, which is annotated with docregions.
To ensure that this file is not ignored by the example-collector, we had to
update the `.gitignore` to not ignore this file. (This also allows the file to be
easily stored tracked in git).

mary-poppins commented:
You can preview 70bdf1a at https://pr43216-70bdf1a.ngbuilds.io/.

mary-poppins commented:
You can preview 10f7cb8 at https://pr43216-10f7cb8.ngbuilds.io/.

mary-poppins commented:
You can preview 2a19278 at https://pr43216-2a19278.ngbuilds.io/.

josmar-crwdstffng reviewed (approved):
The changes meet the requirements for my documentation request.  Thank you.

gkalpak commented on aio/content/examples/i18n/doc-files/nginx:
```diff
@@ -0,0 +1,32 @@
+# #docregion
```
I think `nginx.conf` would be a better filename.

But again, I am not sure whether having these configs as external files offers any benefit over keeping them inline (where they are easier to keep up-to-date).

gkalpak commented on aio/content/examples/i18n/doc-files/rendered-output.html:
```diff
@@ -0,0 +1,4 @@
+<!--#docregion-->
```
This is redundant afaict.

gkalpak commented on aio/tools/transforms/examples-package/services/region-parser.js:
```diff
@@ -58,7 +61,9 @@ module.exports = function regionParser() {
               if (plaster) {
-                region.lines.push(plaster);
+                // Use the same indent as the docregion marker
+                const indent = startRegion[0].split(/[^ ]/,1);
```
Nice one ✨

```suggestion
                const indent = startRegion[0].split(/[^ ]/, 1);
```

gkalpak commented on aio/content/examples/i18n/doc-files/angular.json:
```diff
@@ -0,0 +1,179 @@
+// #docplaster ...
```
I am torn on this one (extracting `angular.json` code snippets from `angular.json` copies).
This is less than ideal, but so is hard-coding code-snippets.

Maybe at least add a comment in [aio/tools/examples/UPDATING.md][1].
(It would be even better if we could have a script that checks to ensure this file is the same as the corresponding boilerplate one (minus the docregion comments), but that might be going a bit too far :grin:)

[1]: https://github.com/angular/angular/blob/2a192782faa15e996be37f6d1fe8afd5b4280d42/aio/tools/examples/UPDATING.md

petebacondarwin replied:
I am going to have a think about fixing this problem of annotating boilerplate files, since there are other cases coming up in other tickets.

petebacondarwin replied:
@gkalpak  - what do you think?

https://github.com/angular/angular/pull/43216/commits/74cf63325e43ec5968104da13883031a29634bca

gkalpak replied:
I am not particularly happy with this approach 😅
(But I probably wouldn't be happy with any approach 😞)

Could you please update [aio/tools/examples/README.md][1] and [aio/tools/examples/UPDATING.md][2] to take this into account? :pray:

[1]: https://github.com/angular/angular/blob/75a3c778b1f7be913f0287423d40fade68ee9adc/aio/tools/examples/README.md#the-example-configjson
[2]: https://github.com/angular/angular/blob/75a3c778b1f7be913f0287423d40fade68ee9adc/aio/tools/examples/UPDATING.md

petebacondarwin replied:
How about this @gkalpak : https://github.com/angular/angular/pull/43216/commits/8b8ad520b86c6a1699a8ceb5e04cb3dd59bf42d0

gkalpak replied:
Looks great! The `list-overrides` command addition is awesome 🏆

gkalpak reviewed (changes requested):
Looks mostly good :+1:
I have some concerns on externalizing specific code types 🤔

petebacondarwin commented on aio/content/examples/i18n/doc-files/commands.sh:
```diff
@@ -0,0 +1,35 @@
+
+# #docregion build-production-french
+ng build --configuration=production,fr
+# #enddocregion build-production-french
```
@josmar-crwdstffng ^^^

josmar-crwdstffng replied:
@gkalpak @petebacondarwin 

There are multiple items that have been brought into scope, but the focus of this PR is to ensure the following.

*   The markdown should always be using code examples that are valid and tested.
*   The markdown should not be the source for code examples.
*   The markdown should be validated using a linter that meets the needs of the docs.

With those items in mind, we may need to a centralized token for commands to ensure the following.

*   commands are valid and tested
*   commands are not running as part of a build

---  

> In response to "We do not lint, test or otherwise validate them (afaict) and it just makes it harder for authors to edit them (and ensure they remain relevant with the surrounding text)."

If an author/editor knows where to look there is no change in difficulty, and any changes are not hidden among markdown changes.
By enforcing that all code examples are placed outside of the markdown, the location of the source code is identified and code is updated when the code changes (not when during an doc update.

We may need a centralized (non-doc specific) location, such as a sample app to pull the doc-region content and ensure the code matches what a reader will experience.

josmar-crwdstffng replied:
@gkalpak @petebacondarwin
As long as the files and doc-regions that are implemented in this PR do not cause issues, we should move forward.

petebacondarwin replied:
> We may need a centralized (non-doc specific) location, such as a sample app to pull the doc-region content and ensure the code matches what a reader will experience.

This is actually what we have at https://github.com/angular/angular/tree/master/aio/content/examples/i18n

The problem is that things like shell commands are things typed in by the user rather than scripts that are executed, so there is no way for these to be verified automatically. But this is true whether they are inline or external. So if you feel that it makes no difference to the doc author then that is fine.

That being said, I think that there are two levels of expertise required. The baseline is being able to find where a doc-region is and to modify it; the more complex requirement is to be able to create these examples (so that they are runnable/testable/etc) and create the doc-regions in the first place. The latter is more tricky for most doc authors I expect.

josmar-crwdstffng replied:
I think we are in agreement.
The doc author should be able to rely on a SME or the tech review (TR) stage to ensure the code example is accurate for the placement in the docs.

gkalpak replied:
I might be missing something, but I don't see the benefit of moving certain types of code-snippets outside the Markdown.

In fact, there is quite an overhead for authors (whomever that might be: the docs team, SWEs and most importantly community contributors). FWIW, our early choice to move the TypeScript and HTML code-snippets outside the Markdown was only made because we believe the benefits of being able to validate the code outweigh the cost of not having the code side-by-side with the rest of the docs content (where it would be much easier to keep in-sync).

So, I strongly believe that there are certain code types (i.e. those we cannot validate) that would be better off being left in the Markdown.

But, since this is related to docs authoring, I am happy to leave it up to the docs team :smiley:

petebacondarwin commented on aio/tools/transforms/examples-package/services/region-parser.js:
```diff
@@ -15,6 +15,8 @@ module.exports = function regionParser() {
     html: html,
     svg: html,
     css: blockC,
+    nginx: inlineHash,
```
👍  - I was basing this off the original example. I'll use `nginx.conf` - thanks

petebacondarwin commented on aio/content/guide/i18n.md:
```diff
@@ -866,7 +866,7 @@ Also, to declare the base `href` at compile time, use the CLI `--baseHref` optio
 Users are redirected to the preferred language defined in the browser using the `Accept-Language` HTTP header.
 If the user has not defined a preferred language, or if the preferred language is not available, then the server falls back to the default language.
-To change the language, see another subdirectory.
+To change the language, see another subdirectory.  
```
Two white spaces (rather than the one that was there before) adds a soft line break, which is what was actually desired but missed from a previous update to the docs.

petebacondarwin replied:
I have reverted the change here, and the change that deleted the single whitespace at the end.
I will leave it to @josmar-crwdstffng to fix this when splitting up this doc.

mary-poppins commented:
You can preview 8686fba at https://pr43216-8686fba.ngbuilds.io/.

mary-poppins commented:
You can preview 57656b4 at https://pr43216-57656b4.ngbuilds.io/.

petebacondarwin committed:
build(docs-infra): support overriding boilerplate files in examples

It is now possible to specify file paths in the `"boilerplate-override"`
property of `example-config.json` files to tell the example boilerplate
system to use the file in the example directory rather than overwriting the
example file with that from the boilerplate directory.

mary-poppins commented:
You can preview eee6ade at https://pr43216-eee6ade.ngbuilds.io/.

petebacondarwin commented on aio/tools/transforms/examples-package/processors/collect-examples.spec.js:
```diff
@@ -205,7 +204,7 @@ describe('collectExampleRegions processor', () => {
 function createDoc(content, relativePath, docType) {
   return {
-    fileInfo: {relativePath: relativePath, extension: path.extname(relativePath).substr(1)},
+    fileInfo: {relativePath: relativePath, baseName: path.basename(relativePath), extension: path.extname(relativePath).substr(1)},
```
Ah that might have been to support a previous change in this PR that we decided was not necessary - i.e. the nginx extensionless config file.

petebacondarwin replied:
I'll remove it

mary-poppins commented:
You can preview 6241cf4 at https://pr43216-6241cf4.ngbuilds.io/.

gkalpak commented on aio/tools/examples/README.md:
```diff
@@ -68,6 +68,7 @@ The file is expected to contain a JSON object with zero or more of the following
   Default: `"cli"`
 - `useCommonBoilerplate: boolean`: Whether to include common boilerplate from the [common/](./shared/boilerplate/common) folder.
   Default: `true`
+- `"override-boilerplate": string[]`: A list of paths to boilerplate files that are overridden by custom files in this example. Commonly this is used when a boilerplate file is referenced in a guide and so needs to have doc-regions added.
```
Can you also mention that the files listed here might need to be "un-ignored" for this specific example (i.e. by adding something like `!<example-name>/path/to/file` in the appropriate place in `content/examples/.gitignore`)? :pray:

gkalpak replied:
Actually, why not use camelCase (easier to access from JS and more consistent with `useCommonBoilerplate`)? :grin:

gkalpak commented on aio/tools/examples/example-boilerplate.js:
```diff
@@ -63,10 +72,39 @@ class ExampleBoilerPlate {
    */
   remove() { shelljs.exec('git clean -xdfq', {cwd: EXAMPLES_BASE_PATH}); }
 
+  listOverrides() {
```
:100:

gkalpak reviewed (approved):
Looks great! :shipit:
The `list-overrides` command addition is awesome 🏆

mary-poppins commented:
You can preview d649db5 at https://pr43216-d649db5.ngbuilds.io/.

mary-poppins commented:
You can preview 094cbfb at https://pr43216-094cbfb.ngbuilds.io/.
